stdlib/ballerina-builtin/src/main/java/org/ballerinalang/nativeimpl/io/utils/IOUtils.java

```diff
@@ -0,0 +1,225 @@
+        WriteBytesEvent writeBytesEvent = new WriteBytesEvent(channel, content, offset, numberOfBytes, context);
+        CompletableFuture<EventResult> future = EventManager.getInstance().publish(writeBytesEvent);
+        EventResult eventResponse = future.get();
+        offset = offset + (Integer) eventResponse.getResponse();
+        Throwable error = ((EventContext) eventResponse.getContext()).getError();
+        if (null != error) {
+            throw new ExecutionException(error);
+        }
+        return offset;
+    }
+
+    /**
+     * <p>
+     * Writes bytes to a channel asynchronously.
+     * </p>
+     *
+     * @param channel       channel which will be used to write bytes.
+     * @param content       content which will be written.
+     * @param offset        the offset which will be set to write bytes.
+     * @param numberOfBytes number of bytes which should be written.
+     * @param context       context of the native function call.
+     * @param function      callback function which should be called upon completion.
+     */
+    public static void write(Channel channel, byte[] content, int offset, int numberOfBytes, EventContext context,
+                             Function<EventResult, EventResult> function) {
+        WriteBytesEvent writeBytesEvent = new WriteBytesEvent(channel, content, offset, numberOfBytes, context);
+        CompletableFuture<EventResult> future = EventManager.getInstance().publish(writeBytesEvent);
+        future.thenApply(function);
+    }
+
+    /**
+     * Asynchronously reads bytes from the channel.
+     *
+     * @param content the initialized array which should be filled with the content.
+     * @param channel the channel the content should be read into.
+     * @param offset  if the array size should be read from an offset
+     * @throws InterruptedException during interrupt error.
+     * @throws ExecutionException   errors which occurs while execution.
+     */
+    public static int readFull(Channel channel, byte[] content, int offset, EventContext context)
+            throws InterruptedException, ExecutionException {
+        int numberOfBytesToRead = content.length - offset;
+        do {
+            offset = offset + read(channel, content, offset, context);
+        } while (offset < numberOfBytesToRead && !channel.hasReachedEnd());
+        return offset;
+    }
+
+    /**
+     * Reads characters from the channel.
+     *
+     * @param characterChannel   channel the characters should be read.
+     * @param numberOfCharacters the number of characters to read.
+     * @param function           the callback function which will be triggered after reading characters.
+     * @param context            context of the event.
+     */
+    public static void read(CharacterChannel characterChannel, int numberOfCharacters,
+                            Function<EventResult, EventResult> function, EventContext context) {
```

Comment:
Unused variable context?

Reply:
Yes. This will be used when async functions are available. Added this since when it's available we could call the function